### PR TITLE
fix: torch.compile correctness for PME and Ewald electrostatics

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

This PR fixes the incorrect `torch.compile` behavior reported in [issue #51](https://github.com/NVIDIA/nvalchemi-toolkit-ops/issues/51), where `particle_mesh_ewald` could produce corrupted compiled forward results and `ewald_summation` could preserve forward values but still return incorrect compiled gradients compared to eager mode.

The root cause was twofold. First, Warp-backed autograd state was not being recovered in a way that reliably survives `torch.compile` / AOT autograd for decorator-backed custom ops, especially when compiled backward sees wrapped tensors or optional tensor inputs. Second, the PME reciprocal-space path hit incorrect behavior in the complex FFT-heavy compiled segment. This change keeps the public electrostatics API intact, fixes the shared `warp_custom_op` compiled-autograd path, and restores eager/compiled parity for the reported electrostatics cases.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes #51

## Changes Made

- `warp_custom_op` now carries Warp backward state through a saved token and replays backward through a registered custom-op path that remains valid under `torch.compile` / AOT autograd.
- The shared backward-state registry now cleans itself up when graphs are abandoned without calling backward, and the generic compiled backward path handles optional tensor inputs correctly.
- Warp launches are bound to PyTorch’s current CUDA stream, and PME uses a narrow compile-only eager boundary around the complex FFT-heavy reciprocal block to avoid incorrect compiled complex-operator lowering.
- Ewald and PME keep only the small compile-specific postprocessing fixes still required for parity, and focused compile regressions plus manual validators cover electrostatics and other decorator-backed Warp ops.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

Focused regression coverage included:

- `python -m pytest -q -p no:cacheprovider -p no:cov -p no:pytest_cov -o addopts='' test/math/bindings/torch/test_autograd.py -k 'torch_compile or compiled_backward_matches_eager or optional_tensor'`
- `python -m pytest -q -p no:cacheprovider -p no:cov -p no:pytest_cov -o addopts='' test/interactions/electrostatics/bindings/torch/test_pme.py -k 'compiled_parity_explicit_params or stream_ordering'`
- `python -m pytest -q -p no:cacheprovider -p no:cov -p no:pytest_cov -o addopts='' test/interactions/electrostatics/bindings/torch/test_ewald.py -k 'compiled_parity'`
- `python nvalchemi_compile_bug.py`
- `python nvalchemi_spline_compile_validate.py`
- `python nvalchemi_coulomb_compile_validate.py`

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

This is a focused correctness fix, not a full architectural solution for compiled PME. The complex FFT-heavy reciprocal block still runs through a narrow eager boundary on compiled paths.

A more rigorous future fix may still be needed if we want the entire PME reciprocal path to remain fully compiler-visible without that eager boundary, especially given TorchInductor's current weak support for complex operators.

The reproducer can still emit a Dynamo graph-break warning if user code calls `estimate_pme_parameters()` directly inside a compiled function, but the incorrect compiled electrostatics results reported in the issue are fixed by this PR.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/torch/autograd.py [hunk 1]
```diff
@@ -256,15 +411,77 @@ def decorator(func: Callable) -> Callable:
...
+                for array_name in grad_arrays
+                if hasattr(first_output, f"_wp_{array_name}")
+            }
+            state = _RegisteredBackwardState(
+                tape=first_output._warp_tape,
+                arrays=arrays,
+            )
+            with _BACKWARD_STATE_LOCK:
+                _BACKWARD_STATE_REGISTRY[token_id] = state
+            return torch.tensor(token_id, dtype=torch.int64)
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/autograd.py]: fixed in a5322c00e025cbd6229a407208161b173e5a0491

File: nvalchemiops/torch/interactions/electrostatics/ewald.py [hunk 2]
```diff
@@ -3059,8 +3068,21 @@ def ewald_summation(
...
+    results = []
+    for idx, (r, s) in enumerate(zip(rs_tuple, rec_tuple)):
+        if (
+            charge_grad_index is not None
+            and idx == charge_grad_index
+            and torch.compiler.is_compiling()
+        ):
+            results.append(_sum_charge_gradients(r, s))
+        else:
+            results.append(r + s)
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/ewald.py]: fixed in a5322c00e025cbd6229a407208161b173e5a0491

**Review by dallasfoster — APPROVED** [review]
